api/src/main/java/org/openmrs/module/openconceptlab/importer/Importer.java

```diff
@@ -113,82 +175,179 @@ public void setZipFile(ZipFile zipFile) {
+							oclConcept.setVersionUrl(prependBaseUrl(baseUrl, oclConcept.getVersionUrl()));
+							oclConcept.setUrl(prependBaseUrl(baseUrl, oclConcept.getUrl()));
+							importTask.setOclConcepts(Arrays.asList(oclConcept));
+						}
+						finally {
+							closeInputStream();
+						}
 
-                if (updatedSince == null) {
-                    oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken());
-					importService.updateReleaseVersion(anImport,
-                            oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken()));
-                } else {
-                    if (subscription.isSubscribedToSnapshot()) {
-                        oclResponse = oclClient.fetchSnapshotUpdates(subscription.getUrl(), subscription.getToken(),
-                                updatedSince);
-                    }
-                    else {
-                        oclResponse = oclClient.fetchOclConcepts(subscription.getUrl(), subscription.getToken(), lastImport.getReleaseVersion());
-						importService.updateReleaseVersion(anImport,
-                                oclClient.getOclReleaseVersion(subscription.getUrl(), subscription.getToken()));
-                    }
-                }
-
-				if (oclResponse != null) {
-					importService.updateOclDateStarted(anImport, oclResponse.getUpdatedTo());
-					importService.updateSubscriptionUrl(anImport, subscription.getUrl());
-					in = new CountingInputStream(oclResponse.getContentStream());
-					totalBytesToProcess = oclResponse.getContentLength();
-
-					processInput();
-
-					in.close();
-				}
-            }
-        });
-    }
+						// Next, read the file into a List of OclMapping and add these to the import task
+						try {
+							openInputStream();
+							JsonParser mappingsParser = objectMapper.getJsonFactory().createJsonParser(in);
+							JsonToken jsonToken = advanceToListOf("mappings", null, mappingsParser);
+							if (jsonToken != JsonToken.END_OBJECT) {
+								List<OclMapping> oclMappings = new ArrayList<>(BATCH_SIZE);
+								while (mappingsParser.nextToken() != JsonToken.END_ARRAY) {
+									OclMapping oclMapping = mappingsParser.readValueAs(OclMapping.class);
+									oclMapping.setUrl(prependBaseUrl(baseUrl, oclMapping.getUrl()));
+									oclMapping.setFromConceptUrl(prependBaseUrl(baseUrl, oclMapping.getFromConceptUrl()));
+									oclMapping.setFromSourceUrl(prependBaseUrl(baseUrl, oclMapping.getFromSourceUrl()));
+									oclMapping.setToConceptUrl(prependBaseUrl(baseUrl, oclMapping.getToConceptUrl()));
+									// We do not import Q-AND-A mappings for which the imported concept is an answer
+									boolean isQuestionAnswer = "Q-AND-A".equals(oclMapping.getMapType());
+									if (isQuestionAnswer && oclMapping.getToConceptUrl().equals(oclConcept.getUrl())) {
+										log.debug("Skipping Q-AND-A mapping for which the imported concept is a mapped answer");
+									}

```

Comment:
Is this a common occurrence? Also, isn't this something that should be fixed outside of this module, i.e., at the concept source itself?

Reply:
Yes, it is common - anytime you are exporting a concept that is the answer to a question.  Look at the export for "TRUE" in OCL - it has 710 "associations", most of which are Q-AND-A mappings for which it is an answer:
https://app.openconceptlab.org/#/orgs/CIEL/sources/CIEL/concepts/1065/

The assumption / decision I made was to not worry about these associations when exporting a concept that is an answer, only worry about those when exporting a question that has this answer.  Make sense?

Reply:
I understand that associations like this are common, but they aren't generally included in the file, depending on how it's generated. E.g., when I download that concept as JSON from the OCL Term Browser, it actually has no mappings in it at all.

[Attachment](https://github.com/openmrs/openmrs-module-openconceptlab/files/7367590/concept-1065.json.txt)

If I download the concept from the API, I can either get the concept with no mappings at:

https://api.openconceptlab.org/orgs/CIEL/sources/CIEL/concepts/1065/

Or with mappings where the concept is the `from` concept:

https://api.openconceptlab.org/orgs/CIEL/sources/CIEL/concepts/1065/?includeMappings=true

Or with mappings where the concept is the `to` concept:

https://api.openconceptlab.org/orgs/CIEL/sources/CIEL/concepts/1065/?includeInverseMappings=true

I can also combine those filter. The point was that, in general, I wouldn't expect someone to end up with a bunch of mappings where the concept is the to concept and if we do we need to guard against it more generally.

Reply:
That's cool that there is that flexibility in the API.  Unfortunately, that same flexibility isn't in the term browser, and the term browser that I see does contain all of those mappings.  This is what I was using to download a Concept.  My experience with OCL is that there is a lot of inconsistency in each of the various pages you might go to in terms of what is included in an export file.  For a Concept like this you get either JSON or ZIP.  For other Collections or Dictionaries, you sometimes get JSON, sometimes CSV, sometimes ZIP formats, and these formats are not necessarily consistent with each other (as you say, sometimes they contain mappings and other details, sometimes they do not).

Reply:
Ah... that's fair. Actually this would be a good thing to put more centrally in mapping exclusions because I suppose the export format could change whether intentionally or accidentally to include inverse mappings.